doc/smart-switch/high-availability/smart-switch-ha-hld.md

```diff
@@ -0,0 +1,2208 @@
+3. The flow replication packet will be sent out with the selected source port in outermost packet, as well as in the metadata. And on the other side, the ack packet can use the same source port too. This binds the packets for this flow to a single data path.
+4. If one of the links went down, all the packets with certain source ports will start to fail. Hence, if we could track the failure for each port in the pool, we can avoid using it.
+5. Alternatively, the DPU-to-DPU probe can use the same ports in the pool, in order to reflect the data path problem. This also helps us to see when the data path is recovered, and we can resume using this port again.
+
+#### 4.2.4. VM to DPU data plane
+
+Again, it is very important to avoid completely killing the VM’s connectivity if one link is blackholing traffic. Hence, when VM sends traffic to DPU, we should also make it leverage ECMP setup.
+
+When VM sends packet out, the packet will be encap’ed by VxLan (see below). Hence, similar to DPU-to-DPU data plane channel, we can also set the VxLan source port by calculating the hash of the inner packet. Then, different flow will go through different data path to reach the SmartSwitch.
+
+<p align="center"><img alt="VM to DPU data plane" src="./images/npu-to-dpu-tunnel-local.svg"></p>
+
+## 5. ENI programming with HA setup
+
+After HA is enabled, instead of programming a single DPU for a single ENI, now we have to program 2 DPUs for a single ENI, so we need to define the behavior of SmartSwitch and its upstream service with HA setup.
+
+### 5.1. Working with upstream service
+
+In SmartSwitch, upstream service and switches will work together on providing the HA features.
+
+The responsibility of these 2 things is defined as below (HA-related only):
+1. Upstream service is responsible for:
+    1. Deciding ENI placement and pairing.
+    2. Deciding preferred active ENI placement.
+    3. Triggering HA operations under planned events, such as planned switchover, planned shutdown for upgrades, evening traffic and etc.
+    4. Triggering HA operations for manual live site mitigations.
+2. SmartSwitch is responsible for:
+    1. Driving the HA state machine transitions.
+    2. Report every ENI HA state change and reasons, so upstream service knows what is happening and can make decisions for planned events.
+    3. Handle HA related requests from upstream service.
+    4. Monitor and handle unplanned events, and trigger defined mitigations, such as driving to standalone setup.
+
+> It is very important to enable SmartSwitch to drive the HA state machine transition on its own. Under unplanned events such as network failures, we could lose connectivity to upstream service. If we could not drive the state machines or make decisions, we would lose the ability to handle the failures.
+
+### 5.2. HA control plane overview
+
+Here is the overview of HA control plane, and we will walk through the components and communication channels in the following sections.
```

Comment:
I still see the redis/zmq channel going over PCIe bus. My understanding was all channels are carried as packets over the traffic lanes.

Reply:
the redis/zmq communication channels will indeed go through the PCIe bus instead of BP ports. the ethernet connection is provided by ethernet-over-pcie.